### PR TITLE
POC for Polymorphism

### PR DESCRIPTION
**Pull request by dabeck81**
POC for Polymorphism

### CODE REVIEW

File: springwolf-core/src/main/java/io/github/springwolf/core/asyncapi/scanners/common/payload/internal/PayloadService.java [hunk 1]
```diff
@@ -49,4 +56,30 @@ public NamedSchemaObject useUnusedPayload() {
...
+        if (messageReferences.isEmpty()) {
+            return null;
+        }
+        if (isConcreteClass(clazz)) {
+            messageReferences = new ArrayList<>(messageReferences);
+            messageReferences.add(MessageReference.toSchema(schemaName));
+        }
+        return SchemaObject.builder()
+                .oneOf(messageReferences.stream()
+                        .sorted(Comparator.comparing(MessageReference::getRef))
```
Comment (timonback) [inline springwolf-core/src/main/java/io/github/springwolf/core/asyncapi/scanners/common/payload/internal/PayloadService.java:76]: side note: Can the ObjectMapper sort the entries at the end?

In earlier versions, we tried sorting entries manually, but over time it became more and more difficult to maintain and debug.

File: springwolf-core/src/main/java/io/github/springwolf/core/asyncapi/scanners/common/payload/NamedSchemaObject.java [hunk 2]
```diff
@@ -1,16 +1,25 @@
...
 
+import io.github.springwolf.asyncapi.v3.model.channel.message.MessageReference;
 import io.github.springwolf.asyncapi.v3.model.schema.SchemaObject;
 import jakarta.annotation.Nullable;
 
 /**
  * Encapsulates the resolved name for the contained schema.
  * @param name The fully qualified name or the simple name of the schema.
  * @param schema The SchemaObject.
+ * @param schemaPayload The schema-payload to be inserted in the message, when not null this schema will override the payload of the message.
```
Comment (timonback) [inline springwolf-core/src/main/java/io/github/springwolf/core/asyncapi/scanners/common/payload/NamedSchemaObject.java:12]: nitpicking: I wonder if this is rather something like a `composedSchema` (composed = anyOf, allOf, oneOf; which right is only used for polymorphic payloads)

Currently the naming represents what it should be used for vs. what it actually is/represents

Comment (timonback) [reply springwolf-core/src/main/java/io/github/springwolf/core/asyncapi/scanners/common/payload/NamedSchemaObject.java:12]: Actually, I just saw that it is called polymorphism in OpenAPI: https://swagger.io/docs/specification/data-models/inheritance-and-polymorphism/
ComposedSchema is used by Swagger, however SchemaObjects are already Springwolf mapped AsyncAPI schemas.
tldr: I am still unsure about the typical CS problem called 'naming things'

File: springwolf-core/src/main/java/io/github/springwolf/core/asyncapi/components/examples/walkers/DefaultSchemaWalker.java [hunk 3]
```diff
@@ -159,6 +159,7 @@ private Optional<T> buildExampleFromUnvisitedSchema(
             String name, Schema schema, Map<String, Schema> definitions, Set<Schema> visited) {
         Optional<Schema<?>> resolvedSchema = resolveSchemaFromRef(schema, definitions);
         if (resolvedSchema.isPresent()) {
+            //            name = exampleValueGenerator.lookupSchemaName(resolvedSchema.get()).orElse(name);
```
[the remark this answers is not in the document]
Comment (dabeck81) [reply springwolf-core/src/main/java/io/github/springwolf/core/asyncapi/components/examples/walkers/DefaultSchemaWalker.java:162]: not needed anymore, removed it

File: springwolf-core/src/main/java/io/github/springwolf/core/asyncapi/schemas/converters/SchemaTitleModelConverter.java [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+@Slf4j
+public class SchemaTitleModelConverter implements ModelConverter {
+    @Override
+    public Schema resolve(AnnotatedType type, ModelConverterContext context, Iterator<ModelConverter> chain) {
+        JavaType javaType = Json.mapper().constructType(type.getType());
+        if (chain.hasNext()) {
+            Schema<?> schema = chain.next().resolve(type, context, chain);
+            if (schema != null && PrimitiveType.createProperty(type.getType()) == null) {
+                if (schema.get$ref() != null) {
+                    Schema<?> defModel = context.resolve(type);
```
[the remark this answers is not in the document]
Comment (dabeck81) [reply springwolf-core/src/main/java/io/github/springwolf/core/asyncapi/schemas/converters/SchemaTitleModelConverter.java:27]: As the method `context.resolve(type)` reads from the swagger map `definedModels`, I called the variable defModel, but definedModel is maybe more intuitive, so I changed it to that

File: springwolf-plugins/springwolf-cloud-stream-plugin/src/test/java/io/github/springwolf/plugins/cloudstream/asyncapi/scanners/channels/CloudStreamFunctionChannelsScannerIntegrationTest.java
[the remark this answers is not in the document]
Comment (dabeck81) [reply springwolf-plugins/springwolf-cloud-stream-plugin/src/test/java/io/github/springwolf/plugins/cloudstream/asyncapi/scanners/channels/CloudStreamFunctionChannelsScannerIntegrationTest.java:1]: yes, that was what I was thinking as well

File: springwolf-examples/springwolf-amqp-example/src/test/resources/asyncapi.json [hunk 5]
```diff
@@ -227,7 +227,7 @@
         },
         "name": "io.github.springwolf.examples.amqp.dtos.AnotherPayloadDto",
         "title": "AnotherPayloadDto",
-        "description": "Another payload model",
+        "description": "",
```
[the remark this answers is not in the document]
Comment (dabeck81) [reply springwolf-examples/springwolf-amqp-example/src/test/resources/asyncapi.json:230]: The description that is shown is the description of the schema, so the logic place to show this description is on the Schema-component.
The description that is on that place, should be the one of the @AsyncMessage, in my opinion.

The reason why the description is not anymore shown, is that Swagger creates now the reference to the schema-object, whereas before it was a hardcoded reference by Springwolf itself. Due to removing that hardcoded reference, we do not receive anymore the full schema, but an already referenced schema-object on the payload, without a description.
If this is a blocking matter, what could be done is:
- Either, fetching the schema from the swagger-reference and fetching the description via that way.
- Either, setting the description on the reference-schema with a custom modelConverter.

File: springwolf-examples/springwolf-kafka-example/src/test/resources/asyncapi.json [hunk 6]
```diff
@@ -25,8 +25,8 @@
     "another-topic": {
       "address": "another-topic",
       "messages": {
-        "io.github.springwolf.examples.kafka.dtos.AnotherPayloadDto": {
-          "$ref": "#/components/messages/io.github.springwolf.examples.kafka.dtos.AnotherPayloadDto"
+        "java.util.ListIo.github.springwolf.examples.kafka.dtos.AnotherPayloadDto": {
```
[the remark this answers is not in the document]
Comment (dabeck81) [reply springwolf-examples/springwolf-kafka-example/src/test/resources/asyncapi.json:28]: hmm, I must admit, that indeed the consumerRecord does not get extracted the way it should. A "modelConverter" is always a good idea to adapt the schema to the way we would like. I think this is a same case as with the MonetaryExample, where Swagger does not see the defined fields of the class

I haven't tried the example on the current code of Springwolf, how it generates. I suppose quite the same.

An idea that comes to my mind is that maybe for each of the different "support-bindings", some custom modelconverters could be added on the classpath. So for example this case: The class `ConsumerRecord` belongs to the Kafka framework, so only when that binding is selected, the support for consumerRecord could be added with a custom modelconverter. You could define that modelconverter bean on the `SpringwolfKafkaBindingAutoConfiguration`.

I suppose each framework has its special quircks, so adding custom modelconverters for each of them to the binding-autoconfiguration seems like the way to go.

I don't know what you think.

File: springwolf-core/src/main/java/io/github/springwolf/core/asyncapi/schemas/SwaggerSchemaService.java [hunk 7]
```diff
@@ -89,12 +99,33 @@ public ExtractedSchemas extractSchema(Class<?> type, String contentType) {
...
+            // defaulting to stringSchema when resolvedSchema is null
+            SchemaObject payloadSchema = swaggerSchemaUtil.mapSchema(
+                    PrimitiveType.fromType(String.class).createProperty());
+            return new Payload(ComponentSchema.of(payloadSchema), Map.of());
+        } else {
+            Map<String, Schema> preProcessSchemas = new LinkedHashMap<>(resolvedSchema.referencedSchemas);
+            Schema payloadSchema = resolvedSchema.schema;
+            preProcessSchemas(payloadSchema, preProcessSchemas, type);
+            HashMap<String, Schema> postProcessSchemas = new HashMap<>(preProcessSchemas);
+            postProcessSchema(preProcessSchemas, postProcessSchemas, actualContentType);
```
[the remark this answers is not in the document]
Comment (dabeck81) [reply springwolf-core/src/main/java/io/github/springwolf/core/asyncapi/schemas/SwaggerSchemaService.java:122]: The duplication is more in the methods, I simply duplicated, like:

postProcessSchema
processAsyncApiPayloadAnnotation
processSchemaAnnotation
processCommonModelConverters

Those 4 methods are twice in the swaggerSchemaService. Once for the old way of defining schemas (still in use for headers) and once for the new way of defining schemas (for defining the payload).
So, as part of a next refactoring the header-generation could be adapted to make also use of "resolvedSchema" and then those duplicate methods can go away.

File: springwolf-examples/springwolf-kafka-example/src/test/resources/asyncapi.json (hunk 6 above)
[the remark this answers is not in the document]
Comment (timonback) [reply springwolf-examples/springwolf-kafka-example/src/test/resources/asyncapi.json]: I added some code as mentioned in https://github.com/springwolf/springwolf-core/issues/891#issuecomment-2335792315

If you don't feel strongly about it, I would like to minimize breaking changes and continue to extract `List`

File: springwolf-examples/springwolf-amqp-example/src/test/resources/asyncapi.json (hunk 5 above)
[the remark this answers is not in the document]
Comment (timonback) [reply springwolf-examples/springwolf-amqp-example/src/test/resources/asyncapi.json:234]: Good point regarding the location of the description and it seems that the description does not belong here.

Springwolf has two ways to specify an optional description via the Swagger Schema annotation (belongs to the schema) and via the AsyncOperation annotation (belongs to the operation).

Can the description be null in that case instead of an empty string?
(I am currently unsure, whether the ui selects the description already from the correct location)
